More details in invalid_attribute_error

jonathanrocher opened the pull request:
Added the bad type received to the error message when attribute lookup fails. To exercise this, you can run:
```
from traits.api import HasTraits, Int

class A(HasTraits):
    x = Int

a = A()
print a.__getattribute__("x")
print a.__getattribute__(0)
```
which would give 
```
0
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/Users/jrocher/Code-samples/Traits/test_getattr_error.py in <module>()
      8 
      9 print a.__getattribute__("x")
---> 10 print a.__getattribute__(0)

TypeError: attribute name must be an instance of <type 'str'>. Got 0 (int).
```

Feedback requested.

jvkersch commented on traits/ctraits.c:
```diff
@@ -310,9 +310,16 @@ fatal_trait_error ( void ) {
+
+    strncat(err_msg, type_name, 20);
+
+    PyErr_SetString( PyExc_TypeError, err_msg );
```
You could use `PyErr_SetObject` and assemble a Python string from a format specifier via `PyString_FromFormat`; that would avoid having to allocate a static buffer array.

jonathanrocher replied:
Thanks @jvkersch for the feedback. I pushed an updated version that uses your suggestion and adds a `repr` of the value for even more feedback. More thoughts?

mdickinson replied:
It sounds as though you actually want `PyErr_Format`.

mdickinson commented:
To avoid printing out error messages of unbounded size, you might want to limit the size of the interpolated strings, by using `%.200s` instead of `%s` (for example).

jonathanrocher commented:
I think I have addressed all comments. There is coveralls unhappy with me because the coverage went down by 0.0%! Not sure what I can do about that...

mdickinson commented:
Great.  If no-one else gets to it first, I'll take a look when I get back from CA.

jonathanrocher commented:
@mdickinson Reminder to take a look...

mdickinson commented:
LGTM.

mdickinson commented on traits/ctraits.c:
```diff
@@ -310,9 +310,27 @@ fatal_trait_error ( void ) {
+#else
+
+    const char* obj_repr_str = PyString_AsString(ob_repr);
+    const char* fmt = "attribute name must be string. Got %.200s (%.200s).";
```
Since `unicode` objects might reasonably be considered to be strings, what do you think about a more specific error message here?  E.g. "attribute name must be an instance of <type 'str'>.  Got ...".

jonathanrocher replied:
Good idea. Done.

mdickinson commented on traits/ctraits.c:
```diff
@@ -310,9 +310,27 @@ fatal_trait_error ( void ) {
+    // Build error message by collecting the type of the object received
+    PyTypeObject* obj_type = name->ob_type;
+    const char* type_name = obj_type->tp_name;
+    PyObject * ob_repr = PyObject_Repr(name);
```
The return value here needs to be checked: if it's `NULL`, that indicates that something went wrong with the `PyObject_Repr` call, and `-1` should be returned.

mdickinson replied:
You've created a new reference here;  there needs to be a corresponding `Py_DECREF` at some point before the end of the function to compensate.

jonathanrocher replied:
Thanks for pointing this out and being explicit. Please take a look and confirm I got it right.

mdickinson commented:
Sorry, I take back the LGTM; there are a couple of minor issues in the C code (reference leaks, error handling) that need addressing.

Please could you also add a changelog entry?

mdickinson commented:
Assigning to Jonathan for updates.

jonathanrocher commented:
I think I have addressed your comments @mdickinson . Thanks for them. Ready for another look.

mdickinson commented:
LGTM

mdickinson commented:
The C code is not valid C89, and the version of MSVC that Appveyor is using doesn't allow the C99 extension of arbitrarily-placed declarations. Fixing ...

mdickinson commented:
@jvkersch, @dpinte: This is probably going to require corresponding changes to #298.

mdickinson commented on traits/ctraits.c:
```diff
@@ -310,10 +310,39 @@ fatal_trait_error ( void ) {
+#if PY_MAJOR_VERSION >= 3
+
+    const char* fmt = "attribute name must be an instance of <type 'str'>. "
+                      "Got %R (%.200s).";
```
This still looks wrong: we're computing the `repr` of `name`, and then showing the `repr` of that `repr`! I'll rework.

mdickinson commented:
@jonathanrocher or @jvkersch: Either of you want to review? If not, I'll merge this in a day or two.

jvkersch commented:
LGTM

mdickinson commented:
@jvkersch: Thanks! Merging.

jonathanrocher commented:
Thanks @mdickinson and @jvkersch for cleaning this up!

mdickinson commented:
@jonathanrocher: Followup question, if you have the time. I seem to recall that the motivation for this PR was an inscrutable error message coming from real code. Do you happen to remember what that real code looked like? (We're currently deciding what to do with this in the Cython version of traits.)

jonathanrocher commented:
@mdickinson Sorry for the delay in response. No I don't remember exactly all the details but I remember that in the real code, adding information led to a method raise a `TypeError` when expecting a string and receiving a unicode. I couldn't make up a simple example of that case, and could only reproduce the error message with the example at the top of the message. 

From the date of the issue, you/I can infer what project I was working on, but I can't remember which part of it gave me that issue...

mdickinson commented:
@jonathanrocher: No problem; thanks for the response!
